CRAMcodecs.tex

```diff
@@ -0,0 +1,3183 @@
+example input.
+
+\begin{minipage}[t]{0.5\textwidth}
+Observed Order-1 frequencies:
+\\[8pt]
+\begin{tabular}{ |r|r|r| }
+\hline
+Context & Symbol & Frequency\\
+\hline
+\textbackslash0 & a & 4 \\
+\hline
+a & a & 3 \\
+  & b & 8 \\
+  & c & 4 \\
+  & d & 4 \\
+\hline
+b & r & 8 \\
+\hline
+c & a & 4 \\
+\hline
+d & a & 4 \\
+\hline
+r & a & 8 \\
+\hline
+\end{tabular}
+\end{minipage}
+\begin{minipage}[t]{0.5\textwidth}
+Normalised (per Order-0 statistics):
+\\[8pt]
+\begin{tabular}{ |r|r|r|}
+\hline
+Context & Symbol & Frequency\\
+\hline
+\textbackslash0 & a & 4095 \\
+\hline
+a & a &  646 \\
+  & b & 1725 \\
+  & c &  862 \\
+  & d &  862 \\
+\hline
+b & r & 4095 \\
+\hline
+c & a & 4095 \\
+\hline
+d & a & 4095 \\
+\hline
+r & a & 4095 \\
+\hline
+\end{tabular}
+\end{minipage}
+
+
+Encoded as:
+\begin{verbatim}
+0x00                 # `\0' context
+0x61      0x8f 0xff  # a  <4095>
+0x00                 # end of Order-0 table
+
+0x61                 # `a' context
+0x61      0x82 0x86  # a            <646>
+0x62 0x02 0x86 0xbd  # b <+2: c,d> <1725>
+          0x83 0x5e  # c (implicit) <862>
+          0x83 0x5e  # d (implicit) <862>
+0x00                 # end of Order-0 table
+
+0x62 0x02            # `b' context, <+2: c, d>
+0x72      0x8f 0xff  # r <4095>
+0x00                 # end of Order-0 table
+
+                     # `c' context (implicit)
+0x61      0x8f 0xff  # a <4095>
+0x00                 # end of Order-0 table
+
+                     # `d' context (implicit)
+0x61      0x8f 0xff  # a <4095>
+0x00                 # end of Order-0 table
+
+0x72                 # `r' context
+0x61      0x8f 0xff  # a <4095>
+0x00                 # end of Order-0 table
+
+0x00                 # end of contexts
+\end{verbatim}
+
+
+\subsection{rANS entropy encoding}
+
+The encoder takes a symbol $s$ and a current state $x$ (initially zero) to
+produce a new state $x'$ with function $C$.
+
+{
+\setlength{\parindent}{1cm}
+\indent $x' = C(s,x)$
+}
+
+The decoding function $D$ is the inverse of $C$ such that $C(D(x)) = x$.
+
+{
+\setlength{\parindent}{1cm}
+\indent $D(x') = (s,x)$
+}
+
+The entire encoded message can be viewed as a series of nested $C$
+operations, with decoding yielding the symbols in reverse order, much
+like popping items off a stack.  This is where the asymmetric part of
+ANS comes from.
+
+As we encode into $x$ the value will grow, so for efficiency we ensure
+that it always fits within known bounds. This is governed by
+
+{
+\setlength{\parindent}{1cm}
+\indent $L \leq x < bL-1$
+}
+
+
+where $b$ is the base and $L$ is the lower-bound.
+
+We ensure this property is true before every use of $C$ and after every
+use of $D$.  Finally to end the stream we flush any remaining data out
+by storing the end state of $x$.
+
+
+\textbf{Implementation specifics}
+
+We use an unsigned 32-bit integer to hold $x$. In encoding it is
+initialised to zero. For decoding it is read little-endian from the
```

Comment:
The current state `x` is initialized to lower bound `L`, not 0.

Reply:
Good catch! Thanks.

Reply:
I also see if wasn't explained in the Nx16 variant, where L is the smaller `1<<15`.  I'll fix that too.